sw/device/silicon_creator/manuf/base/ft_personalize.c

```diff
@@ -587,6 +591,77 @@ static status_t personalize_gen_dice_certificates(ujson_t *uj) {
   return OK_STATUS();
 }
 
+static status_t compute_tbs_was_hmac(perso_blob_t *perso_blob_to_host) {
+  // Read out the WAS from flash.
+  hmac_key_t was;
+  static_assert(
+      kFlashInfoFieldWaferAuthSecretSizeIn32BitWords == kHmacKeyNumWords,
+      "WAS size expected to be same size as HMAC-SHA256 key.");
+  TRY(flash_ctrl_testutils_info_region_setup_properties(
+      &flash_ctrl_state, kFlashInfoFieldWaferAuthSecret.page,
+      kFlashInfoFieldWaferAuthSecret.bank,
+      kFlashInfoFieldWaferAuthSecret.partition, kFlashInfoPage3ReadPermissions,
+      /*offset=*/NULL));
+  TRY(manuf_flash_info_field_read(
+      &flash_ctrl_state, kFlashInfoFieldWaferAuthSecret, was.key,
+      kFlashInfoFieldWaferAuthSecretSizeIn32BitWords));
+
+  // Compute HMAC of TBS certs with WAS as the key.
+  // Most HSMs and host tooling will compute an HMAC in big endian format, so we
+  // do the same to make comparison easier.
+  hmac_hmac_sha256_init(was, /*big_endian_digest=*/true);
+  uint8_t *tlv_buf = perso_blob_to_host->body;
+  uint16_t obj_size;
+  perso_tlv_object_type_t obj_type;
+  perso_tlv_cert_obj_t cert_obj;
+  for (size_t i = 0; i < perso_blob_to_host->num_objs; ++i) {
+    // Parse the TLV object header.
+    perso_tlv_object_header_t obj_header = *(uint16_t *)tlv_buf;
+    PERSO_TLV_GET_FIELD(Objh, Type, obj_header, &obj_type);
+    PERSO_TLV_GET_FIELD(Objh, Size, obj_header, &obj_size);
+    if (obj_type == kPersoObjectTypeX509Tbs) {
+      TRY(perso_tlv_get_cert_obj(tlv_buf, obj_size, &cert_obj));
+      hmac_sha256_update(cert_obj.cert_body_p, cert_obj.cert_body_size);
+    }
+    tlv_buf += obj_size;
+  }
```

Comment:
Can this loop ever complete without finding any `kPersoObjectTypeX509Tbs`?  Should that be an error?

Reply:
good point; fixed.

Reply:
On second thought, if there is no TBS to bind to, it is easier to push nothing into the digest. so we are really just hashing the key and verifying that.